### PR TITLE
Check cursor position for out of bounds of the window

### LINKED ISSUE #8840: `Window.cursor_position` can return `Some` when it shouldn't
## Bevy version

0.10.1

## What you did

Called the `Window.cursor_position` [method](https://docs.rs/bevy_window/0.10.1/bevy_window/struct.Window.html#method.cursor_position) after moving mouse from one window to another.

## What went wrong

It intermittently returns `Some(...)` when it should be `None`.

## Additional information

Using the small app below, this video shows the issue clearly. You can see that sometimes, but not always, when moving the cursor from one window directly to another window, the first window still thinks it has the cursor. This results in multiple windows claiming they have the cursor, while only one can at any one time.

I can see [here](https://github.com/bevyengine/bevy/blob/main/crates/bevy_winit/src/lib.rs#LL474-L481) Bevy sets the cursor position to  `None` when it sees a `WindowEvent::CursorLeft` event from winit, but I verified that it is fired every time, even when the issue occurs, so I don't believe this is a bug in winit.

https://github.com/bevyengine/bevy/assets/7709415/3fb051f9-f5f3-40ab-9d08-3cc98ecc1103

Minimal app:

```rust
use bevy::prelude::*;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins.set(WindowPlugin {
            primary_window: Some(Window {
                position: WindowPosition::At(IVec2::new(400, 200)),
                ..default()
            }),
            ..default()
        }))
        .add_startup_system(setup)
        .add_system(print_cursor_position)
        .run();
}

fn setup(mut commands: Commands) {
    commands.spawn(Window {
        title: "Second window".to_owned(),
        position: WindowPosition::At(IVec2::new(0, 0)),
        ..default()
    });
}

fn print_cursor_position(windows_q: Query<&Window>) {
    for window in windows_q.iter() {
        println!(
            "Window '{}' cursor position: {:?}",
            window.title,
            window.cursor_position()
        );
    }
}
```

### PR DESCRIPTION
**Pull request by Aceeri**
# Objective
Fixes #8840 

Make the cursor position more consistent, right now the cursor position is *sometimes* outside of the window and returns the position and *sometimes* `None`.

Even in the cases where someone might be using that position that is outside of the window, it'll probably require some manual transformations for it to actually be useful.

## Solution
Check the windows width and height for out of bounds positions.

---

## Changelog
- Cursor position is now always `None` when outside of the window.

### CODE REVIEW

**Comment by mockersf** [comment]
It is said in #8840 that the event is correctly fired from winit. If that's the case, could you try and fix the bug where it's not setting the position to None instead of adding bound checks?

**Comment by Plonq** [comment]
Check out my latest comment: https://github.com/bevyengine/bevy/issues/8840#issuecomment-1594371078

File: crates/bevy_window/src/window.rs [hunk 1]
```diff
@@ -266,20 +266,35 @@ impl Window {
...
+                let position = position.as_vec2();
+                if position.x > self.width()
+                    || position.x < 0.0
+                    || position.y > self.height()
+                    || position.y < 0.0
+                {
+                    None
+                } else {
+                    Some(position)
+                }
```
Comment (ickshonpe) [inline crates/bevy_window/src/window.rs:294]: `Rect` could be used here for the bounds check
```suggestion
                let position = position.as_vec2();
                if Rect::new(0., 0., self.width(), self.height()).contains(position) {
                    Some(position)
                } else {
                    None
                }
```

**Comment by Aceeri** [comment]
Tbh I think moving the bounds checking to the setting of physical position would be the best bet here. That way even user setting of physical cursor position doesn't give weird results.

**Comment by Aceeri** [comment]
Actually hmm, maybe doing this on `get` methods is better, this handles problems if the window resizes for whatever reason as well.

**Comment by Plonq** [comment]
Should this maybe be an optional feature? E.g. a boolean argument `ignore_out_of_bounds`, or maybe the inverse?

**Comment by Aceeri** [comment]
> Should this maybe be an optional feature? E.g. a boolean argument `ignore_out_of_bounds`, or maybe the inverse?

I don't think so tbh. I'd be fine with adding a cursor_position_unchecked, but I don't see the use case for it. The oob cursors are highly inconsistent and prob shouldn't be relied on here.

**Review by alice-i-cecile — APPROVED** [review]

**Review by ickshonpe — APPROVED** [review]

**Pull request merged**